src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensStructurelessFileViewProviderRegistrar.java

```diff
@@ -0,0 +1,157 @@
+            ContainerUtil.addAllNotNull(lspFileTypes, languageServerDefinition.getFileTypeMappings().keySet());
+            for (Pair<List<FileNameMatcher>, String> filenameMatcherMapping : languageServerDefinition.getFilenameMatcherMappings()) {
+                ContainerUtil.addAllNotNull(lspFilenameMatchers, filenameMatcherMapping.getFirst());
+            }
+        }
+
+        // Determine the abstract file types for those file types and filename matchers
+        Set<FileType> registerFileTypes = new LinkedHashSet<>();
+        for (FileType fileType : FileTypeManager.getInstance().getRegisteredFileTypes()) {
+            if (fileType instanceof AbstractFileType abstractFileType) {
+                if (lspFileTypes.contains(abstractFileType)) {
+                    registerFileTypes.add(abstractFileType);
+                } else {
+                    List<FileNameMatcher> filenameMatchers = FileTypeManager.getInstance().getAssociations(abstractFileType);
+                    if (ContainerUtil.intersects(lspFilenameMatchers, filenameMatchers)) {
+                        registerFileTypes.add(abstractFileType);
+                    }
+                }
+            }
+        }
+
+        // Use working copies to avoid potential dirty reads
+        Set<FileType> copyOfRegisteredAbstractFileTypes = new HashSet<>(registeredAbstractFileTypes);
+        Set<FileType> workingRegisteredAbstractFileTypes = new HashSet<>();
+
+        // Register file types to use our file view provider factory if necessary
+        if (!registerFileTypes.isEmpty()) {
+            for (FileType registerFileType : registerFileTypes) {
+                FileViewProviderFactory currentFactory = FileTypeFileViewProviders.INSTANCE.findSingle(registerFileType);
+                if (!(currentFactory instanceof LSPSemanticTokensStructurelessFileViewProviderFactory)) {
+                    FileTypeFileViewProviders.INSTANCE.addExplicitExtension(registerFileType, LSPSemanticTokensStructurelessFileViewProviderFactory.INSTANCE);
+                }
+                workingRegisteredAbstractFileTypes.add(registerFileType);
+            }
+        }
+
+        // Unregister any obsolete registrations
+        Collection<FileType> unregisterFileTypes = ContainerUtil.subtract(copyOfRegisteredAbstractFileTypes, workingRegisteredAbstractFileTypes);
+        if (!unregisterFileTypes.isEmpty()) {
+            for (FileType unregisterFileType : unregisterFileTypes) {
+                FileTypeFileViewProviders.INSTANCE.removeExplicitExtension(unregisterFileType, LSPSemanticTokensStructurelessFileViewProviderFactory.INSTANCE);
+            }
+        }
+
+        // If changed, update our retained file type/factory associations
+        if (!copyOfRegisteredAbstractFileTypes.equals(workingRegisteredAbstractFileTypes)) {
+            registeredAbstractFileTypes = workingRegisteredAbstractFileTypes;
+        }
+    }
+
+    // LISTENERS
```

Comment:
All of these listeners pretty much just call the registration logic above which will be a no-op if no changes are required.